### ISSUE TITLE
[Docos] Scientific examples or @beartype didn't happen

### ISSUE THREAD

**Issue by NLPShenanigans** [beartype/beartype#224]
Hey @leycec, I must say I am a roaring fan of beartype and all the labyrinthine twists and turn that have come with the scintillating stochastic fun that is the Bare Metal Bear :) While I've not had a chance to contribute more development-leaning issues as of yet, one area that would get more folks involved and raise the bearhype would be to have a couple more end-to-end examples of data analytic or data scientific workflows in the Bearpedia. For instance:

- For data processing pipelines where dynamic-type checks can be incorporated as a part of vetting data quality, one can read in some raw data via Pandas or Polars, build some custom type annotations to semantically encode for what a data analyst thinks 'good' data should look like, apply the beartype decorator to test for entries that don't adhere to the 'goodness' expectations, let the boolean mask be a new column, save the data to a parquet format, and read-in the reformatted data using the type-checking boolean mask for predicate pushdown. This way, you're out-of-the-gate only looking at data this is 'sensible', according to agreed-upon typing conventions.
- In a similar flavour, given that working with Excel, outside of the narrowest of fields, is an utter train-wreck (case in point: https://www.theverge.com/2020/8/6/21355674/human-genes-rename-microsoft-excel-misreading-dates), and yet so ubiquitous for those in the business world as to be wholly unavoidable, a flushed out example or two using a combination of ftfy (for those that are sick and tired of mojibake), beartype, and phantom-types, for a wonderful blend of addressing encoding malformations, argument, and value checks to try to unmuck the muck with some ft-as-a-fydle, rawrcous, phantasmal fun!
- For data scientific workflows, having a toy example using a couple canonical datasets with some common ML frameworks (Pytorch, Tensorflow, Pytorch Lightning, Equinox, or the like) leveraging a combination of jaxtyping and numerary type hints (since PyTrees for days are the saving grace we never knew we needed), could be instructive.
- For those more towards data reliability engineering, I see beartype working well in parallel to ideate around data contract schema (something in the flavour of soda-core, where non-technical users can give their salt alongside we more numerically-nutty folk).

These thoughts are rather rough as I'm only one espresso deep, and I could stand to brainstorm with others as the best work was seldom build either in isolation or in a day, but I would be very interested in contributing more here, working towards a collaborative publication to Medium perhaps with folks from the community when we have something we're collectively happy with, notwithstanding wanting to just get more involved and share in some interesting conversations with folks :)

**Comment by leycec** [comment]
You're so far ahead of the type-checking curve here you've catapulted off into uncharted territory. Let's hope this is a Moebius Strip. With any luck and more than a little hand-waving, let's also hope that @beartype can do everything you want without me doing anything more than nodding and stroking my handlebar moustache.

> I am a roaring fan of beartype

![Conan bear](https://media.tenor.com/d44R4tqRxqgAAAAS/conan-conan-bear.gif)

> the scintillating stochastic fun that is the Bare Metal Bear

![RNGesus please](https://media.tenor.com/o3f0gb89FAUAAAAC/thorobr-rng.gif)

> This way, you're out-of-the-gate only looking at data this is 'sensible', according to agreed-upon typing conventions.

**OMG!** That sounds *amazing.* I'm stupefied that I haven't done this already. Simply stupefied. Sadly, I also have no idea how to do this. That probably explains the prior two sentences.

I'd :heart_eyes: to add a new FAQ entry or example showcase exhibiting your proposed type-checking **+** data science pipeline. But I don't know how, because I'm deficient on the data science side. I am dumb is what I'm saying. [Beartype validators](https://beartype.readthedocs.io/en/latest/api_vale) are almost certainly involved somewhere here, but...

I'm unclear exactly how. The [`beartype.vale.Is[...]` factory](https://beartype.readthedocs.io/en/latest/api_vale/#beartype.vale.Is) enables Turing-complete (i.e., arbitrary) decisions at type-checking time. In theory, we *should* be able to stuff exactly what you want into a sufficiently non-trivial `beartype.vale.Is[...]` validator that dynamically validates the contents of the passed Pandas- and/or Polars-backed `DataFrame` according to some user-defined quality metrics for satisfiability. In practice, I continue stroking my handlebar moustache.

Please, somebody! If it will help, I will sacrifice an Elmo doll on a graven altar for this.

> given that working with Excel, outside of the narrowest of fields, is an utter train-wreck

**Yes.** Please don't do that, anybody. When you use Excel, you are officially in the Outer Limits of data science. @beartype can no longer help you there. Our powers only reach so far.

> mojibake

**OMG.** Somebody wrote Japanese on our issue tracker! Everybody, cluster around! Something profoundly exciting that has never happened before just happened! 

I see your mojibake and raise you 文字化け.

> using a combination of ftfy (for those that are sick and tired of mojibake), beartype, and phantom-types, for a wonderful blend of addressing encoding malformations, argument, and value checks to try to unmuck the muck with some ft-as-a-fydle, rawrcous, phantasmal fun!

Pretty sure my neocortex just lit up like a bioluminescent flower.

Seriously, though. Yet again, I can only nod and look hopefully at the growing audience of NLP specialists hovering over @beartype's shoulder. Having just glanced at [the `ftfy` API](https://ftfy.readthedocs.io/en/latest/fixes.html), it's unclear whether `ftfy` currently publishes the sort of *detection* functions that @beartype would need to correctly detect invalid (or just poorly formatted) Unicode sequences; `ftfy` only appears to publish various high- and low-level *transformation* functions (e.g., `ftfy.fix_text()`, `ftfy.fixes.decode_inconsistent_utf8()`).

Presumably, `ftfy` already internally defines those detection functions. After all, `ftfy` itself needs to detect BadDudes™ in Unicode, right? Perhaps we can kindly prod them to publish something enabling @beartype integration.

If you find a spare moment lying around,<sup>so... *never*</sup> would you mind posing this same question to the [`ftfy` devs on their GitHub issue tracker](https://github.com/rspeer/python-ftfy)? @beartype can finish this gauntlet – but the journey of a thousand ~~knives~~ UTF-8 digraphs probably begins in `ftfy`'s backyard.

> a couple canonical datasets with some common ML frameworks (Pytorch, Tensorflow, Pytorch Lightning, Equinox, or the like) leveraging a combination of jaxtyping and numerary type hints (since PyTrees for days are the saving grace we never knew we needed), could be instructive.

**Yes!** @patrick-kidger, I proudly summon you. You are the stalwart typing-centric data guru needed in this moment. Since somebody mentioned `jaxtyping`, I'm now handing off the reins to you. I bet you're about to recuse yourself, right? Wait! Patrick, wait! We just wonder if you might have any example `jaxtyping` **+** @beartype **+** ML framework **+** open-source dataset lying around.

...no? Nobody? Yeah. Kinda didn't think so. I am making the exhaling face. :face_exhaling: 

> For those more towards data reliability engineering, I see beartype working well in parallel to ideate around data contract schema (something in the flavour of soda-core, where non-technical users can give their salt alongside we more numerically-nutty folk).

I... have never even heard of those nouns before. I suspect I'm not supposed to admit that, but I will. When you live in a cabin in the Canadian wilderness, you rapidly learn to swallow your pride.

In hindsight, it makes sense that data reliability is a thing and that highly-paid hyper-specialists would be remunerated in literal gold bars to do that thing *vis-a-vis* data contract schemas, Soda Core, and other profitable nouns I glibly pretend to understand while stroking a beard.

In short, **excitement intensifies.** Let's do something by 2047 about this, everybody!

**Comment by antonagestam** [comment]
Let me know if I can help out from the phantom-types side of things. Would be great to see more content produced that sheds light on these tools 👍

**Comment by leycec** [comment]
**@antonagestam spittin' truth.** `phantom-types` rocks the typing Casbah and enables so much of what we do here at @beartype. Real-world examples integrating both would almost be a miraculous act of divinity at this point.

I'll happily merge *anything* into our docos that resembles a working solution. Go, Typonistas! Go!

**Comment by patrick-kidger** [comment]
FWIW [Equinox](https://github.com/patrick-kidger/equinox) itself uses beartype during its tests: [see here](https://github.com/patrick-kidger/equinox/blob/5298e154eb1146d5f49f87a408b51f3ff042f401/pytest.ini#L2).

That's not super obvious though. I agree it'd be nice to have an example somewhere that makes explicit use of jaxtyping and beartype. I'd be happy to host that in the Equinox docs, or to point people at such an example hosted in the beartype docs.

I don't really have time to put something together myself, but I'd be happy to shephard a PR or offer feedback on a blog post that does this.

**Comment by NLPShenanigans** [comment]
Hey @leycec, your purple flair has once again left me gutted 🤣 Handlebar mustachio twirling notwithstanding, I have some demo script showcasing this already that is sitting woefully neglected, and would be happy to add this to a new FAQ once I do a little house-keeping to make it appropriate for public-viewing (i.e., tidy), so thankfully no Elmo sacrifices required 😁

Yes, Excel is like the Outer Limits; strange goings-ons are liable to go bump, irrespective of the time of day or night 😅 I've been raised! As I know not a lick of Kanji, touché! Haha high praise indeed, especially as bioluminescence on the brain is, I've been told, the bee's knees!

Insofar as ftfy is concerned, you are correct that it's logic for handling Unicode malformations is, by design, quite heuristic. This has to do with the fact that the subset of the combinatorial space of all possible character sequences which deals with common Unicode-malformed sequences is contextualized by language, and especially for short sequences (such as with abbreviations), any stochastic techniques, machine learning or no, would perform dreadfully.

Taking my last remark at face value, consider the word Götamål, a dialect within the Swedish language family. While a regex match on sequences like må in Swedish is a valid character sequence, but in English, a regex match on må would be a valid heuristic for Unicode malformation, requiring a map or a sequence of maps that would high likelihood correctly address the malformation. The encoding identification portion is a minefield (see https://ftfy.readthedocs.io/en/latest/detect.html), so if you a web developer prior to Unicode standardization decided to be sadistic and leave out the metadata for their choice of encoding on their web page, tough luck 😅

With this in mind, a gentle message to the maintainers to work on language flagging as part of the transformation functions could help make an integration vaguely plausible, worthwhile as it is. Yeah I'll happily raise it with them in the meantime that there's interest to do so (whenever I can time-dilate my way into having enough time in the day to structure a reasonable request) 😁

Yup, though I dare say they'd even go with bars of Latinum 😂 If anyone has more experience on the data reliability engineering end of things, please announce yourself and let's chat!

Thanks @antonagestam  and @patrick-kidger for the speedy and, as-per-usual, succinct replies 😀 I admit I literally discovered phantom-types by perusing beartype issues, as one does, I will want to take time to familiarize myself with the code base a bit more before I would say I'd produce examples that anyone would consider constructive, let alone instructive, so any soundboarding @antonagestam would be tremendously helpful 😀 @patrick-kidger, oh yes quite right! I entirely missed that 😂 Knowledge dropped! I'll happily receive some shepherding and/or feedback once I flush out something worth looking at 😀 Happy times for the Bare Metal Bear and company!

**Comment by leycec** [comment]
**Gah!** I somehow missed your wondrously delicious reply, @NLPShenanigans. GitHub waste products incessantly chatter in my inbox. I quietly ignore them. This is the price! So... *let's see.*

@beartype now [officially supports pandera type hints](https://github.com/beartype/beartype/issues/227) – a popular third-party solution for runtime type-checking pandas data frames, indices, and series. Tests say this work. I partially believe them.

@beartype now also documents this support [via a **new FAQ entry**](https://beartype.readthedocs.io/en/latest/faq/#pandas-data-frames). It is whimsical. Yet, I can only scrunch up my face with fatigue. In theory, this addresses a sizeable chunk of your lead question:

> For data processing pipelines where dynamic-type checks can be incorporated as a part of vetting data quality, one can read in some raw data via Pandas or Polars, build some custom type annotations to semantically encode for what a data analyst thinks 'good' data should look like, apply the beartype decorator to test for entries that don't adhere to the 'goodness' expectation...

Yes! Yes! So much yes! @beartype now does all that *vis-a-vis* pandera. Thank pandera. You may also thank us by paying us money. Swedish krona is acceptable.

I'm kidding! I jest. I'm a fanciful jester. We do this for the love of chatbots that invariably insult my "unique" communication style and poor cross-country skiing technique.

> I have some demo script showcasing this already that is sitting woefully neglected, and would be happy to add this to a new FAQ...

**OMG!** Arise from the ashes of history, Swedish Empire! @beartype will fan you with grape leaves soaked in Canadian maple syrup if you lend us your beautiful examples. Let us FAQ them! FAQ them all, I say.

> Götamål, a dialect within the Swedish language family.

...*what is this linguistic madness.* I thought there was just Swedish! Like, isn't Swedish this indivisible monolithic thing? Now, I learn the answer is: "No. There are many Swedish languages, which all disagree with one another. And you well never learn *any* of them, @leycec. Feel bad."

Also, I read the word "Götamål" while envisioning a burly man with a bloodstained beard blowing on a horn that reaches all the way to the next ice-wracked village.

> This has to do with the fact that the subset of the combinatorial space of all possible character sequences which deals with common Unicode-malformed sequences is contextualized by language, and especially for short sequences (such as with abbreviations), any stochastic techniques, machine learning or no, would perform dreadfully.

I can no longer feel my brain. I think this means you just blew my mind.

> With this in mind, a gentle message to the maintainers to work on language flagging as part of the transformation functions could help make an integration vaguely plausible, worthwhile as it is.

Are you quite certain you're Swedish? Your erudite diction, prolix nebulosity, and witty repartee suggests you are, in fact, a Victorian-era poet named Gotamal.

> Yup, though I dare say they'd even go with bars of Latinum

...no. It can't be! You can't have just dropped a subtle *Star Trek* reference with a glib half-smirk that you thought nobody would notice. But you did. I'm wise to your superior intellect now, @NLPShenanigans. Clearly, you have publicly exposed yourself as a time traveller from our post-capitalist space-faring future. *Ian Banks was right about everything, wasn't he!?!?*

`</ahem>`

We have now left the normal plenum of acceptable behaviour. This can only mean one thing... **let's close this issue!** Hah, hah. Didn't see that one coming, did you? Our new FAQ entry sort of satisfies rudimentary concerns. Also, @leycec is tired and now speaking about himself in the third person. If you unearth any *awesome* examples, please drop us another issue, PR, or discussion thread and we'll happily merge everything.

Thanks again for the fecund debate, everybody. NLP + @beartype = the world wins. :grin:
